src/test/java/org/kiwiproject/registry/consul/server/ConsulRegistryServiceIntegrationTest.java

```diff
@@ -0,0 +1,199 @@
+        void shouldRegisterAndReturnRegisteredInstance_OverridingDomainIfConfiguredToDoSo() {
+            var now = Instant.now();
+            when(environment.currentInstant()).thenReturn(now);
+
+            var serviceInstance = ServiceInstance.fromServiceInfo(ServiceInfoHelper.buildTestServiceinfoWithHostName("example.com"))
+                    .withStatus(ServiceInstance.Status.UP);
+            config.setDomainOverride("test");
+
+            service.register(serviceInstance);
+
+            var services = consul.catalogClient().getService(serviceInstance.getServiceName()).getResponse();
+
+            var hostName = services.stream().map(CatalogService::getServiceAddress).findFirst();
+            assertThat(hostName).isPresent().get().isEqualTo("example.test");
+        }
+    }
+
+    @Nested
+    class UpdateStatus {
+
+        @Test
+        void shouldReturnNullIfNotRegistered() {
+            assertThat(service.updateStatus(ServiceInstance.Status.UP)).isNull();
+        }
+
+        @Test
+        void shouldReturnRegisteredInstanceUntouchedWhenRegistered() {
+            var serviceInstance = ServiceInstance.builder().build();
+            service.registeredService.set(serviceInstance);
+            assertThat(service.updateStatus(ServiceInstance.Status.UP)).isEqualToComparingFieldByField(serviceInstance);
+        }
+    }
+
+    @Nested
+    class UnRegister {
+
+        @Test
+        void shouldUnRegister() {
+            service.registeredService.set(ServiceInstance.builder().serviceName("APPID").instanceId("INSTANCEID").build());
+            consul.agentClient().register(ImmutableRegistration.builder().name("APPID").id("INSTANCEID").address("localhost").build());
+
+            service.unregister();
+
+            assertThat(service.registeredService.get()).isNull();
+            assertThat(consul.catalogClient().getService("APPID").getResponse()).isEmpty();
+        }
+
+        @Test
+        void shouldThrowIllegalStateExceptionWhenNeverRegistered() {
+            assertThatThrownBy(service::unregister)
+                    .isInstanceOf(IllegalStateException.class)
+                    .hasMessage("Cannot unregister since registration was never called");
+        }
+
+        @Test
+        void shouldRetryUnregisterAndThrowExceptionIfAllTriesExpire() {
+            var badConsul = mock(Consul.class);
+            var serviceToFail = new ConsulRegistryService(badConsul, new ConsulRegistrationConfig(), environment);
+            serviceToFail.registeredService.set(ServiceInstance.builder().serviceName("APPID").instanceId("INSTANCEID").build());
+
+            doThrow(new RuntimeException("Oops")).when(badConsul).agentClient();
+
+            assertThatThrownBy(serviceToFail::unregister)
+                    .isInstanceOf(RegistrationException.class)
+                    .hasMessage("Error un-registering service APPID, id INSTANCEID");
+
+            assertThat(serviceToFail.registeredService.get()).isNotNull();
```

Comment:
If we added that method `getRegisteredService` it could be package private and used in all these cases...